### Practical 2 - Philippe Noël

In [38]:
import pandas as pd
import numpy as np
from sklearn.metrics import log_loss
from sklearn.feature_extraction.text import CountVectorizer
from IPython.display import display

In [39]:
#load data
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

#separate the response 
train_response = train[['toxic','severe_toxic','obscene','threat','insult','identity_hate']]
train_input = train[['comment_text']]

#countvectorizer
cv = CountVectorizer()
x_train = cv.fit_transform(train.comment_text.values)
x_test = cv.transform(test.comment_text.values)

In [40]:
#display the head of the training data
train.head()

Unnamed: 0            id  \
0       11846  122236411367   
1       12413  127962107102   
2       67989  708200507036   
3        2452   25036827254   
4       86438  902006105668   

                                        comment_text  toxic  severe_toxic  \
0  Vandalism\nReverting to Macedonian ethnicity w...      0             0   
1  "\nok, I've put all his uploads there, removed...      0             0   
2  "\n\n Paul LePage \nStop messaging me, as I re...      0             0   
3  Fraudulent Book ???\nHow is the book fraudulen...      0             0   
4  "Given his behaviours, I am starting to wonder...      0             0   

   obscene  threat  insult  identity_hate  
0        0       0       0              0  
1        0       0       0              0  
2        0       0       0              0  
3        0       0       0              0  
4        0       0       0              0

In [47]:
from sklearn.model_selection import train_test_split

# splitting the training data into train & test so we have some values to check against and gage the rate of success
x_train2, x_test2, y_train2, y_test2 = train_test_split(train_input, train_response, test_size=0.20)

#countvectorizer for these new subdatasets -- for the algorithm fitting formats to work
x_train2 = cv.fit_transform(x_train2.comment_text.values)
x_test2 = cv.transform(x_test2.comment_text.values)

In [58]:
# we decided upon a gradient boosting classifier for its similarity to random forest, yet strong performance
from sklearn.ensemble import GradientBoostingClassifier

# output var
train_predict = pd.DataFrame(columns = train_response.columns)

#loop through the response and fit a classifier for each one
for col in train_response.columns:
    # fit for this column
    GBC = GradientBoostingClassifier()
    GBC.fit(x_train2, y_train2[col].values)
    train_predict[col] = GBC.predict_proba(x_test2)[:,1]
    
# display our prediction
display(train_predict)

toxic  severe_toxic   obscene    threat    insult  identity_hate
0     0.064919      0.002397  0.022412  0.000845  0.026417       0.001572
1     0.037860      0.002397  0.011729  0.000845  0.014197       0.001572
2     0.050118      0.002397  0.030459  0.000845  0.028347       0.001572
3     0.074478      0.002397  0.030459  0.000845  0.031995       0.001572
4     0.037496      0.002397  0.011729  0.000845  0.014624       0.001572
5     0.160110      0.002397  0.028694  0.000845  0.054018       0.001572
6     0.046429      0.002397  0.015971  0.000845  0.018414       0.001572
7     0.062446      0.002397  0.021756  0.000845  0.031732       0.001572
8     0.065603      0.002397  0.016967  0.000845  0.020558       0.001572
9     0.024453      0.002397  0.010885  0.000845  0.012572       0.001572
10    0.092207      0.002397  0.020485  0.000845  0.030213       0.001572
11    0.044282      0.002397  0.023800  0.000845  0.027476       0.001572
12    0.140499      0.002397  0.028694  0.000845  0.054018       0.001572
13    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
14    0.026780      0.002397  0.010885  0.000845  0.012951       0.001572
15    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
16    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
17    0.037860      0.002397  0.011729  0.000845  0.014197       0.001572
18    0.171082      0.002397  0.030459  0.000845  0.048981       0.001572
19    0.076764      0.002397  0.030459  0.000845  0.031995       0.001572
20    0.030148      0.002397  0.013982  0.000845  0.013782       0.001572
21    0.057436      0.002397  0.015752  0.000845  0.020558       0.001572
22    0.053530      0.002397  0.021756  0.000845  0.021201       0.001572
23    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
24    0.271691      0.002397  0.160722  0.000845  0.018750       0.001572
25    0.037643      0.002397  0.015063  0.000845  0.014761       0.001572
26    0.068505      0.002397  0.020485  0.000845  0.030213       0.001572
27    0.043173      0.002397  0.013982  0.000845  0.019157       0.001572
28    0.037852      0.002397  0.022107  0.000845  0.028945       0.001572
29    0.039176      0.002397  0.012464  0.000845  0.014924       0.001572
...        ...           ...       ...       ...       ...            ...
9555  0.028757      0.002397  0.011568  0.000845  0.011683       0.001572
9556  0.902559      0.387186  0.856166  0.000845  0.731873       0.012052
9557  0.045166      0.002397  0.010885  0.000845  0.012761       0.001572
9558  0.025745      0.002397  0.012464  0.000845  0.013195       0.001572
9559  0.125736      0.003021  0.105780  0.000845  0.029008       0.001572
9560  0.036156      0.002397  0.015752  0.000845  0.017120       0.001572
9561  0.056374      0.002397  0.023800  0.000845  0.033149       0.001572
9562  0.034181      0.002397  0.011729  0.000845  0.012761       0.001572
9563  0.042402      0.002397  0.010885  0.000845  0.014297       0.001572
9564  0.024558      0.002397  0.012464  0.000845  0.013195       0.001572
9565  0.028185      0.002397  0.012464  0.000845  0.013934       0.001572
9566  0.057095      0.002397  0.020485  0.000845  0.020175       0.001572
9567  0.033410      0.002397  0.015063  0.000845  0.013782       0.001572
9568  0.058294      0.002397  0.023800  0.000845  0.033149       0.001572
9569  0.052112      0.002397  0.010885  0.000845  0.014297       0.001572
9570  0.036686      0.002397  0.016003  0.000845  0.015273       0.001572
9571  0.050859      0.002397  0.014826  0.000845  0.018561       0.001572
9572  0.071910      0.002397  0.028306  0.000845  0.031995       0.001572
9573  0.068746      0.002397  0.023800  0.000845  0.029234       0.001572
9574  0.051787      0.002397  0.022107  0.000845  0.033149       0.001572
9575  0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
9576  0.941097      0.043112  0.946843  0.000845  0.748966       0.002532
9577  0.03474

In [60]:
from sklearn.metrics import log_loss

# testing for log loss accuracy of the training subset
ll_score = log_loss(y_test2, train_predict)
display(ll_score) # 0.325 not bad, let's compare with the random forst

0.32529175501275326

In [59]:
# baseline random forest code for log loss comparison on modified input set
from sklearn.ensemble import RandomForestClassifier

train_predict2 = pd.DataFrame(columns = train_response.columns)

#loop through the response and fit a classifier for each one
for col in train_response.columns:
    rf = RandomForestClassifier()
    rf.fit(x_train2, y_train2[col].values)
    train_predict2[col] = rf.predict_proba(x_test2)[:,1]
    
# error
ll_score2 = log_loss(y_test2, train_predict2)
display(ll_score2) # as we can see, our fitting is a lot better --> mission accomplished

0.980462473441427

Clearly, 0.325 is a lot better than 0.980 so mission accomplished, we beat the baseline random forest. While this is not directly <0.199, we infer since the random forest has 0.199 on the full set that this will have much better than 0.199 log loss error on the full set (which we can't test remotely). We now run this better fit algorithm on our actual test sample, refitting with the whole dataset for even better accuracy.

In [61]:
# running our better-fit algorithm on the actual test set

# real output var
test_predict = pd.DataFrame(columns = train_response.columns)

#loop through the response and fit a classifier for each one
for col in train_response.columns:
    # fit for this column
    GBC = GradientBoostingClassifier()
    GBC.fit(x_train, train_response[col].values)
    test_predict[col] = GBC.predict_proba(x_test)[:,1]
    
# display our prediction
display(train_predict)

toxic  severe_toxic   obscene    threat    insult  identity_hate
0     0.064919      0.002397  0.022412  0.000845  0.026417       0.001572
1     0.037860      0.002397  0.011729  0.000845  0.014197       0.001572
2     0.050118      0.002397  0.030459  0.000845  0.028347       0.001572
3     0.074478      0.002397  0.030459  0.000845  0.031995       0.001572
4     0.037496      0.002397  0.011729  0.000845  0.014624       0.001572
5     0.160110      0.002397  0.028694  0.000845  0.054018       0.001572
6     0.046429      0.002397  0.015971  0.000845  0.018414       0.001572
7     0.062446      0.002397  0.021756  0.000845  0.031732       0.001572
8     0.065603      0.002397  0.016967  0.000845  0.020558       0.001572
9     0.024453      0.002397  0.010885  0.000845  0.012572       0.001572
10    0.092207      0.002397  0.020485  0.000845  0.030213       0.001572
11    0.044282      0.002397  0.023800  0.000845  0.027476       0.001572
12    0.140499      0.002397  0.028694  0.000845  0.054018       0.001572
13    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
14    0.026780      0.002397  0.010885  0.000845  0.012951       0.001572
15    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
16    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
17    0.037860      0.002397  0.011729  0.000845  0.014197       0.001572
18    0.171082      0.002397  0.030459  0.000845  0.048981       0.001572
19    0.076764      0.002397  0.030459  0.000845  0.031995       0.001572
20    0.030148      0.002397  0.013982  0.000845  0.013782       0.001572
21    0.057436      0.002397  0.015752  0.000845  0.020558       0.001572
22    0.053530      0.002397  0.021756  0.000845  0.021201       0.001572
23    0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
24    0.271691      0.002397  0.160722  0.000845  0.018750       0.001572
25    0.037643      0.002397  0.015063  0.000845  0.014761       0.001572
26    0.068505      0.002397  0.020485  0.000845  0.030213       0.001572
27    0.043173      0.002397  0.013982  0.000845  0.019157       0.001572
28    0.037852      0.002397  0.022107  0.000845  0.028945       0.001572
29    0.039176      0.002397  0.012464  0.000845  0.014924       0.001572
...        ...           ...       ...       ...       ...            ...
9555  0.028757      0.002397  0.011568  0.000845  0.011683       0.001572
9556  0.902559      0.387186  0.856166  0.000845  0.731873       0.012052
9557  0.045166      0.002397  0.010885  0.000845  0.012761       0.001572
9558  0.025745      0.002397  0.012464  0.000845  0.013195       0.001572
9559  0.125736      0.003021  0.105780  0.000845  0.029008       0.001572
9560  0.036156      0.002397  0.015752  0.000845  0.017120       0.001572
9561  0.056374      0.002397  0.023800  0.000845  0.033149       0.001572
9562  0.034181      0.002397  0.011729  0.000845  0.012761       0.001572
9563  0.042402      0.002397  0.010885  0.000845  0.014297       0.001572
9564  0.024558      0.002397  0.012464  0.000845  0.013195       0.001572
9565  0.028185      0.002397  0.012464  0.000845  0.013934       0.001572
9566  0.057095      0.002397  0.020485  0.000845  0.020175       0.001572
9567  0.033410      0.002397  0.015063  0.000845  0.013782       0.001572
9568  0.058294      0.002397  0.023800  0.000845  0.033149       0.001572
9569  0.052112      0.002397  0.010885  0.000845  0.014297       0.001572
9570  0.036686      0.002397  0.016003  0.000845  0.015273       0.001572
9571  0.050859      0.002397  0.014826  0.000845  0.018561       0.001572
9572  0.071910      0.002397  0.028306  0.000845  0.031995       0.001572
9573  0.068746      0.002397  0.023800  0.000845  0.029234       0.001572
9574  0.051787      0.002397  0.022107  0.000845  0.033149       0.001572
9575  0.079321      0.002397  0.030459  0.000845  0.031995       0.001572
9576  0.941097      0.043112  0.946843  0.000845  0.748966       0.002532
9577  0.03474

In [62]:
# generating output
test_predict.to_csv('practical2_PN_output.csv')